Fix flaky TestRoutingDataCache due to test method order issue

If testCurrentStatesSelectiveUpdate() runs before testUpdateOnNotification() and then testSelectiveUpdates(). testSelectiveUpdates() would fail.

The reason is that testCurrentStatesSelectiveUpdate() stop one participant.  Thus the code in testSelectiveUpdate() constructing RoutingDataCache, the external view of "TestDb" would be version 1. Note, due to liveInstance change (by stopping one participant), controller would calculate new best possible and update external view of TestDB after converging. Normally, this new best possible external view would shows later than constructing RoutingDataCache in testSelectiveUpdate() or construction of RoutingDataCache in testUpdateOnNotification. Then the cluster converge for losing one participant and external view of TestDB got written to version 2. Thus any `cache.refresh(accessor);` in these two test methods may see external view for TestDB version changed. Thus, it would reload it from ZK even it was cached. This would make the counter greater than expected.

As following two log shows:

> [ERROR] Failures: 
[ERROR]   TestRoutingDataCache.testSelectiveUpdates:103 expected:<1> but was:<2>
[INFO] 
[ERROR] Tests run: 4, Failures: 1, Errors: 0, Skipped: 0

>[ERROR] testSelectiveUpdates(org.apache.helix.spectator.TestRoutingDataCache)  Time elapsed: 0.007 s  <<< FAILURE!
java.lang.AssertionError: expected:<0> but was:<1>
	at org.apache.helix.spectator.TestRoutingDataCache.testSelectiveUpdates(TestRoutingDataCache.java:87)



`Assert.assertEquals(accessor.getReadCount(PropertyType.EXTERNALVIEW),` would fail due to version 2 not match existing version 1 in the cache. 

We need to ensure testCurrentStatesSelectiveUpdate always run later.

```
  public void testSelectiveUpdates()
      throws Exception {
    MockZkHelixDataAccessor accessor =
        new MockZkHelixDataAccessor(CLUSTER_NAME, new ZkBaseDataAccessor<ZNRecord>(_gZkClient));

    RoutingDataCache cache =
        new RoutingDataCache("CLUSTER_" + TestHelper.getTestClassName(), PropertyType.EXTERNALVIEW);
    cache.refresh(accessor);
    Assert.assertEquals(accessor.getReadCount(PropertyType.EXTERNALVIEW), 1);

    accessor.clearReadCounters();

    // refresh again should read nothing
    cache.refresh(accessor);
    Assert.assertEquals(accessor.getReadCount(PropertyType.EXTERNALVIEW), 0);

    // refresh again should read nothing
    cache.notifyDataChange(HelixConstants.ChangeType.EXTERNAL_VIEW);
    cache.refresh(accessor);
    Assert.assertEquals(accessor.getReadCount(PropertyType.EXTERNALVIEW), 0);

    // add new resources
    _gSetupTool.addResourceToCluster(CLUSTER_NAME, "TestDB_1", 1, STATE_MODEL);
    _gSetupTool.rebalanceStorageCluster(CLUSTER_NAME, "TestDB_1", _replica);

    Thread.sleep(100);
    ZkHelixClusterVerifier _clusterVerifier =
        new BestPossibleExternalViewVerifier.Builder(CLUSTER_NAME).setZkClient(_gZkClient).build();
    Assert.assertTrue(_clusterVerifier.verifyByPolling());

    accessor.clearReadCounters();

    // refresh again should read only new current states and new idealstate
    cache.notifyDataChange(HelixConstants.ChangeType.EXTERNAL_VIEW);
    cache.refresh(accessor);
    Assert.assertEquals(accessor.getReadCount(PropertyType.EXTERNALVIEW),

```

--- a/helix-core/src/test/java/org/apache/helix/spectator/TestRoutingDataCache.java
+++ b/helix-core/src/test/java/org/apache/helix/spectator/TestRoutingDataCache.java
@@ -39,6 +39,8 @@ public class TestRoutingDataCache extends ZkStandAloneCMTestBase {
 
   @Test
   public void testUpdateOnNotification() {
+    Assert.assertTrue(_clusterVerifier.verifyByPolling());
+    
     MockZkHelixDataAccessor accessor =
         new MockZkHelixDataAccessor(CLUSTER_NAME, new ZkBaseDataAccessor<ZNRecord>(_gZkClient));
 
@@ -63,6 +65,15 @@ public class TestRoutingDataCache extends ZkStandAloneCMTestBase {
   @Test(dependsOnMethods = { "testUpdateOnNotification" })
   public void testSelectiveUpdates()
       throws Exception {
+    // Added verifier to make sure the test starts at a stable state. Note, if
+    // testCurrentStatesSelectiveUpdate() run first. This test may fail without
+    // this line. The reason is that when testCurrentStatesSelectiveUpdate()
+    // stop one participant, this would trigger liveInstance update in controller
+    // which would lead to new external view for TestDB get updated. The update is
+    // async to the construction of RoutingDataCache in this test and subsequent
+    // refresh().
+    Assert.assertTrue(_clusterVerifier.verifyByPolling());
+
     MockZkHelixDataAccessor accessor =
         new MockZkHelixDataAccessor(CLUSTER_NAME, new ZkBaseDataAccessor<ZNRecord>(_gZkClient));
 
@@ -87,8 +98,6 @@ public class TestRoutingDataCache extends ZkStandAloneCMTestBase {
     _gSetupTool.rebalanceStorageCluster(CLUSTER_NAME, "TestDB_1", _replica);
 
     Thread.sleep(100);
-    ZkHelixClusterVerifier _clusterVerifier =
-        new BestPossibleExternalViewVerifier.Builder(CLUSTER_NAME).setZkClient(_gZkClient).build();
     Assert.assertTrue(_clusterVerifier.verifyByPolling());
 
     accessor.clearReadCounters();
@@ -106,7 +115,6 @@ public class TestRoutingDataCache extends ZkStandAloneCMTestBase {
     _gSetupTool.addResourceToCluster(CLUSTER_NAME, "TestDB_3", 1, STATE_MODEL);
     _gSetupTool.rebalanceStorageCluster(CLUSTER_NAME, "TestDB_3", _replica);
 
-    Thread.sleep(100);
     Assert.assertTrue(_clusterVerifier.verifyByPolling());
 
     // Totally four resources. Two of them are newly added.
@@ -119,7 +127,6 @@ public class TestRoutingDataCache extends ZkStandAloneCMTestBase {
 
     _gSetupTool.getClusterManagementTool().enableResource(CLUSTER_NAME, "TestDB_2", false);
 
-    Thread.sleep(100);
     Assert.assertTrue(_clusterVerifier.verifyByPolling());
 
     cache.notifyDataChange(HelixConstants.ChangeType.EXTERNAL_VIEW);

